Sources/NIOWebSocket/WebSocketOpcode.swift

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+//
+//  WebSocketOpcode.swift
+//  NIOWebSocket
+//
+//  Created by Cory Benfield on 16/02/2018.
+//
+
+/// An operation code for a websocket frame.
+public enum WebSocketOpcode {
+    case continuation
+    case text
+    case binary
+    case unknownNonControl(UInt8)
+    case connectionClose
+    case ping
+    case pong
+    case unknownControl(UInt8)
+
+    /// Create an opcode from the encoded representation.
+    ///
+    /// - parameters
+    ///     - encoded: The encoded representation of the opcode as an 8-bit integer.
+    ///          Must be no more than 4 bits large.
+    public init?(_ encoded: UInt8) {
+        guard encoded < 0x10 else {
+            return nil
+        }
+
+        switch encoded {
+        case 0x0:
+            self = .continuation
+        case 0x1:
+            self =  .text
+        case 0x2:
+            self = .binary
+        case 0x8:
+            self = .connectionClose
+        case 0x9:
+            self = .ping
+        case 0xA:
+            self = .pong
+        case let oc where (oc & 0x8) != 0:
+            self = .unknownControl(oc)
+        default:
+            self = .unknownNonControl(encoded)
+        }
+    }
+
+    /// Convert this opcode to an 8-bit integer, with the opcode in the four
+    /// least-significant bits in a form suitable for sending on the wire. Is
+    /// `nil` if the opcode is not actually a valid websocket opcode.
+    public var asInt: UInt8? {
```

Comment:
Any particular reason this wasn't:

```swift
public extension UInt8 {
  public init?(_ opcode: WebSocketOpcode) {
    ...
  }
}
```

You did create an extension on `UInt16`  for `WebSocketErrorCode` which I think is nice.

Reply:
@ianpartridge Honestly I don't know. I have been trying to get accustomed to the Swift pattern of adding initialisers, rather than converters (my python brain doesn't remember that I can do that), so in all honesty I probably just forgot. Will update.

Reply:
Cool, it's no big deal of course.  Overall this PR is a massive +1!

Have you tested with https://github.com/crossbario/autobahn-testsuite - it's really useful.

Reply:
@Lukasa @ianpartridge agree.... we should run some automated tests with the test suite at some point. We also use it in netty to verify our WebSockets implementation